crypto/bio/bio_mem.c

```diff
@@ -98,56 +105,96 @@ BIO *BIO_new_mem_buf(const void *buf, ossl_ssize_t len) {
 }
 
 static int mem_new(BIO *bio) {
-  BUF_MEM *b;
+  BIO_BUF_MEM *bbm = OPENSSL_zalloc(sizeof(*bbm));
+
+  if (bbm == NULL) {
+    return 0;
+  }
 
-  b = BUF_MEM_new();
-  if (b == NULL) {
+  bbm->buf = BUF_MEM_new();
+  if (bbm->buf == NULL) {
+    OPENSSL_free(bbm);
     return 0;
   }
 
   // |shutdown| is used to store the close flag: whether the BIO has ownership
   // of the BUF_MEM.
+  bbm->read_off = 0;
   bio->shutdown = 1;
   bio->init = 1;
   bio->num = -1;
-  bio->ptr = (char *)b;
+  bio->ptr = (char *)bbm;
 
   return 1;
 }
 
 static int mem_free(BIO *bio) {
+  BIO_BUF_MEM *bbm;
+
+  bbm = (BIO_BUF_MEM *)bio->ptr;
+  BUF_MEM *b = bbm->buf;
+
   if (!bio->shutdown || !bio->init || bio->ptr == NULL) {
     return 1;
   }
 
-  BUF_MEM *b = (BUF_MEM *)bio->ptr;
   if (bio->flags & BIO_FLAGS_MEM_RDONLY) {
     b->data = NULL;
   }
   BUF_MEM_free(b);
   bio->ptr = NULL;
+
+  OPENSSL_free(bbm);
   return 1;
 }
 
+static int mem_buf_sync(BIO *bio) {
```

Comment:
* This function should return void. Caller never check the return value (and this function always returns 0 anyways).
* Is the impact of a failure of this function that the caller would be unable to append new data after the `b->length` offset?

Reply:
> Is the impact of a failure of this function that the caller would be unable to append new data after the b->length offset?

This function shouldn't fail though? If the buffer is empty/null or is read-only, there's nothing for `mem_buf_sync` to sync. Otherwise, `memmove` would be executed, which shouldn't fail either. If the caller wants to append new data, it already has checks to make sure the buffer is neither null nor read-only